src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -42,51 +67,66 @@ pub struct ObligationForest<O, T> {
     /// at a higher index than its parent. This is needed by the
     /// backtrace iterator (which uses `split_at`).
     nodes: Vec<Node<O>>,
-    trees: Vec<Tree<T>>,
-    snapshots: Vec<usize>,
+    /// A cache of predicates that have been successfully completed.
+    done_cache: FnvHashSet<O::Predicate>,
+    /// An cache of the nodes in `nodes`, indexed by predicate.
```

Comment:
Is this a *cache* -- or just a map? I tend to think of cache as meaning "a subset of the full set, recording some previous result"